src/Middleware/StaticFiles/src/StaticFilesEndpointRouteBuilderExtensions.cs

```diff
@@ -0,0 +1,219 @@
+        }
+
+        /// <summary>
+        /// Adds a specialized <see cref="RouteEndpoint"/> to the <see cref="IEndpointRouteBuilder"/> that will match
+        /// requests for non-filenames with the lowest possible priority. The request will be routed a
+        /// <see cref="StaticFileMiddleware"/> that attempts to serve the file specified by <paramref name="filePath"/>.
+        /// </summary>
+        /// <param name="builder">The <see cref="IEndpointRouteBuilder"/>.</param>\
+        /// <param name="pattern">The route pattern.</param>
+        /// <param name="filePath">The file path of the file to serve.</param>
+        /// <param name="options"><see cref="StaticFileOptions"/> for the <see cref="StaticFileMiddleware"/>.</param>
+        /// <returns>The <see cref="IEndpointRouteBuilder"/></returns>
+        /// <remarks>
+        /// <para>
+        /// <see cref="MapFallbackToFile(IEndpointRouteBuilder, string, string, StaticFileOptions)"/> is intended to handle
+        /// cases where URL path of the request does not contain a filename, and no other endpoint has matched. This is
+        /// convenient for routing requests for dynamic content to a SPA framework, while also allowing requests for
+        /// non-existent files to result in an HTTP 404.
+        /// </para>
+        /// <para>
+        /// The order of the registered endpoint will be <c>int.MaxValue</c>.
+        /// </para>
+        /// <para>
+        /// This overload will use the provided <paramref name="pattern"/> verbatim. Use the <c>:nonfile</c> route contraint
+        /// to exclude requests for static files.
+        /// </para>
+        /// </remarks>
+        public static IEndpointConventionBuilder MapFallbackToFile(
+            this IEndpointRouteBuilder builder,
+            string pattern,
+            string filePath,
+            StaticFileOptions options)
+        {
+            if (builder == null)
+            {
+                throw new ArgumentNullException(nameof(builder));
+            }
+
+            if (pattern == null)
+            {
+                throw new ArgumentNullException(nameof(filePath));
+            }
+
+            if (filePath == null)
+            {
+                throw new ArgumentNullException(nameof(filePath));
+            }
+
+            return builder.MapFallback(pattern, CreateRequestDelegate(builder, filePath, options));
+        }
+
+        private static RequestDelegate CreateRequestDelegate(
+            IEndpointRouteBuilder builder,
+            string filePath,
+            StaticFileOptions options = null)
+        {
+            var app = builder.CreateApplicationBuilder();
+            app.Use(next => context =>
+            {
+                context.Request.Path = "/" + filePath;
+
+                return next(context);
+            });
+
+            if (options == null)
+            {
+                app.UseStaticFiles();
```

Comment:
The main things you get from the static file middleware are an IFileProvider, content-type resolution, and conditionals support (if-modified-since...).

The file provider can be assumed (IWebHostEnvironment.WebRootFileProvider), content-types have defaults or the user can specify the content-type if they're hard-coding the path, and it's not clear conditionals would be useful for fallback scenarios.

I don't think you need the full middleware to handle this. Directly calling one of the HttpResponse.[SendFileAsync](https://github.com/aspnet/AspNetCore/blob/master/src/Http/Http.Extensions/src/SendFileResponseExtensions.cs) overloads seems adequate for this scenario. 

E.g.
- `MapFallbackToFile(... filePath)` content-types is resolved from a default set and the file is resolved from IWebHostEnvironment.WebRootFileProvider
- `MapFallbackToFile(... filePath, contentType)` is resolved from IWebHostEnvironment.WebRootFileProvider
- `MapFallbackToFile(... filePath, contentType fileProvider)` is resolved from the given provider
- `MapFallbackToFile(... iFileInfo, contentType)` sends the given file directly

Yet again, if we rewrote static files as endpoints this wouldn't be an issue, you'd new up a pre-resolved static file endpoint.

Reply:
@SteveSandersonMS can elaborate more than I can on the scenarios. The existing SPA stuff and the people that use it use quite a lot of those features. 

In particular caching headers, and multiple file providers have come up in our discussions. It would be really nice if as part of this experience users could configure these things in one place and have them work for both static files and this. 

-----

I'm still open to the idea that this would be something different from the static files middleware, but what's the drawback of using the middleware?

Reply:
The strongest reason we had to use `UseStaticFiles` in SPA-land was to get all the ETag stuff for free. You really have to support this, and `HttpResponse.SendFileAsync` won't give it to you.

I agree it would be nice if the relevant parts of `UseStaticFiles` were factored out, but I also feel (like @rynowak) that there's no obvious drawback in just reusing `UseStaticFiles` here as-is.

Reply:
Discussed offline and we're going to keep this as-is for now. We're meeting after preview 4 to design another approach for static files in general.